lib/sles4sap/console_redirection.pm

```diff
@@ -43,33 +43,133 @@ sub handle_login_prompt {
+    $args{ssh_user} //= get_required_var('REDIRECT_TARGET_USER');
+    $args{ssh_tunnel_port} //= '22022';
+
+    croak 'Package autossh is not installed' if script_run('rpm -qi autossh');
+
+    record_info('Redirection init', 'Preparing console redirection');
+
     # This should get worker VM id before any redirection happening
     # ID serves as identification of the 'base' VM to return to.
-    set_var('WORKER_VM_ID', script_output 'cat /etc/machine-id');
+    set_var('BASE_VM_ID', script_output 'cat /etc/machine-id');
+
+    # Prepare keyless access from remote host to worker VM
+    connect_target_to_serial(%args);
+    script_run("sudo rm $reverse_ssh_key_base_name*", quiet => 1);
+    # remove any existing entry in known_hosts file
+    script_run("ssh-keygen -R [localhost]:$args{ssh_tunnel_port} -f ~/.ssh/known_hosts", quiet => 1);
+    script_run("sudo ssh-keygen -R [localhost]:$args{ssh_tunnel_port} -f /root/.ssh/known_hosts", quiet => 1);
+    assert_script_run("ssh-keygen -f $reverse_ssh_key_base_name -t rsa -b 2048 -N ''");
+    my $public_key = script_output("cat $reverse_ssh_key_base_name.pub", quiet => 1);
+    disconnect_target_from_serial(%args);
+
+    # Add cloud VM key to worker VM authorized keys
+    assert_script_run("echo \"$public_key\" >> ~/.ssh/authorized_keys", quiet => 1);
+
+    # Forward common ports.
+    # # Starts permanent reverse SSH connection from remote VM to worker VM.
+    remote_port_forward(destination_port => $args{ssh_tunnel_port},
+        ssh_user => $args{ssh_user},
+        destination_ip => 'localhost',
+        monitor_port => '20000');
+
+    # Port below is also required to access resources using generated port
+    remote_port_forward(destination_port => get_var("QEMUPORT") + 1,
+        ssh_user => $args{ssh_user},
+        destination_ip => get_var('QEMU_HOST_IP', '10.0.2.2'),
+        monitor_port => '20001');
+    record_info('Redirected', 'Console redirection ready.');
+}
+
+=head2 remote_port_forward
+
+    remote_port_forward(source_port=>$source_port, destination_ip=>$destination_ip
+        [, destination_port=>$destination_port, monitor_port=>$monitor_port, source_ip=>$source_ip]);
+
+B<source_ip>: Source IP address or hostname to forward incomming traffic from. Default: REDIRECT_TARGET_IP
+
+B<destination_ip>: Destination IP or hostname where will the traffic be forwarded. Can be localhost as well.
+
+B<monitor_port>: Port for autossh to monitor tunnel status. Needs to be unique for each autossh instance. Default: off
+
+B<source_port>: Local port to forward traffic from. Default:
+
+B<destination_port>: remote port that local port should be forwarded to, default: local_port=remote_port
+
+B<ssh_user>: Login user for B<source_ip> host.
+
+Forwards traffic from B<source_ip>:B<source_port> to B<destination_ip>:B<destination_port>.
+For example it allows to fetch or upload logs from/to OpenQA instance from cloud host directly instead of copying through jumphost.
```

Comment:
`OpenQA instance` , `cloud host`, `jumphost`

Reply:
Is it better now: 
```
For example: It allows Cloud based SUT uploading logs directly to openQA instance without being able to resolve it directly. 
```